src/SFA.DAS.AssessorService.Application/Handlers/Certificates/GetCertificatesHistoryHandler.cs

```diff
@@ -0,0 +1,118 @@
+
+        public GetCertificatesHistoryHandler(ICertificateRepository certificateRepository,
+            IContactQueryRepository contactQueryRepository,
+            IAssessmentOrgsApiClient assessmentOrgsApiClient,
+            ILogger<GetCertificatesHistoryHandler> logger)
+        {
+            _certificateRepository = certificateRepository;
+            _contactQueryRepository = contactQueryRepository;
+            _assessmentOrgsApiClient = assessmentOrgsApiClient;
+            _logger = logger;
+        }
+
+        public async Task<PaginatedList<CertificateHistoryResponse>> Handle(GetCertificateHistoryRequest request, CancellationToken cancellationToken)
+        {
+            const int pageSize = 10;
+
+            var certificates = await _certificateRepository.GetCertificateHistory(
+                request.Username,
+                request.PageIndex ?? 1,
+                pageSize);
+
+            // Please Note:- Cannot seem to automap this with custom value/type converters
+            // so dealing with it manually for now.
+            var certificateHistoryResponses = MapCertificates(certificates);
+
+            return certificateHistoryResponses;
+        }
+
+        private PaginatedList<CertificateHistoryResponse> MapCertificates(PaginatedList<Certificate> certificates)
+        {
+            var trainingProviderName = string.Empty;
+            var recordedBy = string.Empty;
+            var certificateResponses = certificates.Items.Select(
+                certificate =>
+                {
+                    var certificateData = JsonConvert.DeserializeObject<CertificateData>(certificate.CertificateData);
+                    recordedBy = certificate.CertificateLogs
+                        .FirstOrDefault(certificateLog => certificateLog.Status == Domain.Consts.CertificateStatus.Submitted)?.Username;
+
+                    try
+                    {
+                        if (certificate.Organisation.EndPointAssessorUkprn.HasValue)
+                        {
+                            if (certificateData.ProviderName == null)
+                            {
+                                var provider = _assessmentOrgsApiClient
+                                    .GetProvider(certificate.ProviderUkPrn).GetAwaiter()
+                                    .GetResult();
+                                trainingProviderName = provider.ProviderName;
+                            }
+                            else
+                            {
+                                trainingProviderName = certificateData.ProviderName;
+                            }
+                        }
+                    }
+                    catch (EntityNotFoundException e)
+                    {
+                        _logger.LogInformation(
+                            $"Cannot find training provider for ukprn {certificate.Organisation.EndPointAssessorUkprn.Value}");
+                    }
```

Comment:
As per the plan, can you update the CertificateData (in the db) with the ProviderName once you've got it from the Api please?  Makes it 'self-healing'.

Reply:
What's the need for the try - catch? 😄

Reply:
One more thing, why is this checking for certificate.Organisation.EndPointAssessorUkprn.HasValue?

Reply:
Discussed and dealt with -